runtime/lua/vim/iter.lua

```diff
@@ -179,6 +199,51 @@ function ListIter.filter(self, f)
   return self
 end
 
+--- Flattens a |list-iterator| pipeline, un-nesting nested values.
+--- The flattened list will exclude nil, and any values with non-numeric keys
+---
+--- Example:
+---
+--- ```lua
+--- vim.iter({1, {2}, {{3}}}):flatten():totable()
+--- ```
+--- (returns `{1, 2, {3}}`)
+---
+--- ```lua
+--- vim.iter({1, {a = 2}, {nil, {nil}}, {3}}):flatten():totable()
+--- ```
+--- (returns `{1, {}, 3}`)
```

Comment:
Why does it not fully flatten the table? Because the default `depth` is 1? What would be the rationale of choosing 1 as the default depth? Shouldn't we flatten the entire table by default?

Reply:
By choosing that default, I was trying to mimic the behavior of the [ECMAScript method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) `Arrays.prototype.flat`, which was mentioned in the original issue thread. That is probably the implementation with which people are most familiar. 

Fully flattening by default is definitely another option here. ES uses a depth of infinity to fully flatten a table, but comparisons with infinity, which could be less intuitive to users. in Lua: 

```lua
-- flatten a table to infinite depth (math.huge = inf)
-- alternatively could use 1 / 0
print(vim.inspect(vim.iter.flatten({1, {2}, {{3}}, {{{{4}}}}}, math.huge)))

```

Reply:
I see, that makes a sense and I'm now convinced with `depth` defaulting to 1. New (better) defaults for the new API is good :)